extensions/amp-base-carousel/1.0/base-carousel.js

```diff
@@ -112,16 +153,34 @@ function BaseCarouselWithRef(
     (currentSlide + dir < 0 ||
       (!mixedLength && currentSlide + visibleCount + dir > length));
 
-  const [hadTouch, setHadTouch] = useState(false);
+  const [interaction, setInteraction] = useState(Interaction.NONE);
   const hideControls = useMemo(() => {
     if (controls === Controls.ALWAYS || outsetArrows) {
       return false;
     }
     if (controls === Controls.NEVER) {
       return true;
     }
-    return hadTouch;
-  }, [hadTouch, controls, outsetArrows]);
+    return interaction === Interaction.TOUCH;
+  }, [interaction, controls, outsetArrows]);
+
+  const debouncedAdvance = useMemo(
+    () =>
+      debounce(
+        window,
+        (currentSlide) => {
+          advance(autoAdvanceCount, currentSlide);
+        },
+        autoAdvanceInterval
+      ),
+    [autoAdvanceCount, autoAdvanceInterval, advance]
+  );
+  useEffect(() => {
+    if (interaction !== Interaction.NONE || !shouldAutoAdvance.current) {
+      return;
+    }
+    debouncedAdvance(currentSlide);
```

Comment:
So, we running advance from rendering side effect? Feels a bit dangerous even with debouncing, but might be ok. But is there any other way to trigger it? E.g. from the previous `autoAdvance` action? Or from the `onSlideChanged`?

Reply:
FMI, why is this particularly dangerous?

We can trigger again from `advance`, however the interaction-based stopping (on focus, hover, or touch), can be one transition delayed because the second call is already scheduled at the time the first call occurs. 

We can alternatively trigger from `onSlideChanged` (in the call to `setRestingIndex`) but this adds an additional delay to the base `1000ms` because the slide state is itself updated on debounced scrolling.

I'd prefer the latter over the former alternative from the end user perspective, and we can adjust the given interval by the scroll debounce interval if needed. But it would help me to better understand the motivations for avoiding the current approach.

Reply:
We could also use `win.setInterval`?

Reply:
Let me think about this for a second. But one note: if we used `setInterval`, it'd nicely fit with:

```
useEffect(() => {
  if (!autoAdvance) {
     return;
  }
  const interval = win.setInterval(() => {
    ...
  }, TIMEOUT);
  return () => win.clearInterval(interval);
}, [autoAdvance]);
```

Though, I don't know if this solves all your use cases. But I thought it looked a bit nicer since it's more of declarative.

Reply:
Done. I also had the `autoAdvance` return a bool for success or not and use that to clear the interval in the callback.